improve error handling

lovasoa opened the pull request:
fixes https://github.com/rust-lang/stacker/issues/115

See https://github.com/sqlpage/SQLPage/issues/814

nagisa commented on src/lib.rs:
```diff
@@ -411,31 +404,58 @@ cfg_if! {
+            let pthread_ctx = attr.as_mut_ptr();
+
+            // Linux and BSD use different functions to get attributes of created thread
+            cfg_if! {
+                if #[cfg(any(target_os = "freebsd", target_os = "dragonfly", target_os = "illumos"))] {
+                    let get_attr = libc::pthread_attr_get_np;
+                } else {
+                    let get_attr = libc::pthread_getattr_np;
+                }
+            };
```
Given how compact `guess_os_stack_limit` was before, merging the implementations together only to special case the implementation in the middle makes this part of the change roughly net-negative in my eyes.

`handle_pthread_err` handling resource lifetime(s) is also somewhat iffy -- at this point I'd reach for introducing a wrapper with a destructor for `pthread_attr_t`. Maybe that'll help to clean up the conditional here as well (by e.g. hiding it away into a method?)

If nothing else I would like this to be structured such that the code would not compile if the block of conditions above was modified, but not here, if only to force future maintainer(s) to check this area of code twice.

nagisa replied:
(Stylistically if the lifetime handling of `pthread_attr_t` is moved to be destructor based, then conversions from return code to `Some`/`None` and early returning on error become a fairly straightforward `(res != 0).then_some(())?;`.

nagisa reviewed (commented):
I think it'd be fine to return `None` from `guess_os_stack` when `procfs` isn't available, but in principle it also doesn't change that for most intents and purposes systems configured that way are sorta broken for the purposes of `*-linux-gnu` targets at least.

lovasoa commented:
@nagisa I reorganized guess_os_stack to have separate files per backend, addressed your comments, and added more error handling where it was missing on other platforms.

lovasoa commented:
Great, thank you very much @nagisa !
